### PR TITLE
Show meaningful error when a node is started with invalid settings instead of a guice error

### PR DESCRIPTION
**Pull request by seut**
Accessing the ClusterService while it was not initialized correctly due to an (settings-) error leads to a confusing Guice error instead of the correct setting-is-invalid error.

This is solved by registering ClusterStateListener only onStart of LifecyleComponents
to be sure that the ClusterService is accessible already 

- all SchemaInfo impl. won't listen to the cluster state directly now but 
  instead are called by the Schemas cluster state listener
- default `doc` DocSchemaInfo will be registered by a cluster state event
- all remaining cluster state listener are converted to LifecyleComponents

### CODE REVIEW

**Comment by seut** [comment]
@dobe pls review

File: core/src/main/java/io/crate/ClusterIdService.java [hunk 1]
```diff
@@ -26,41 +26,43 @@
...
 import org.elasticsearch.common.logging.Loggers;
 import org.elasticsearch.common.settings.Settings;
 
 import static org.elasticsearch.cluster.ClusterState.builder;
 
-public class ClusterIdService implements ClusterStateListener {
+@Singleton
+public class ClusterIdService extends AbstractLifecycleComponent<ClusterIdService> implements ClusterStateListener {
+
+    private static final ESLogger LOGGER = Loggers.getLogger(ClusterIdService.class);
```
Comment (dobe) [inline core/src/main/java/io/crate/ClusterIdService.java]: use logger field

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 2]
```diff
@@ -1,12 +1,12 @@
 /*
- * Licensed to Crate.IO GmbH ("Crate") under one or more contributor
+ * Licensed to CRATE Technology GmbH ("Crate") under one or more contributor
```
Comment (dobe) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: i think this is an old header

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 3]
```diff
@@ -21,25 +21,244 @@
...
-    String DEFAULT_SCHEMA_NAME = "doc";
+@Singleton
+public class Schemas extends AbstractLifecycleComponent<Schemas> implements Iterable<SchemaInfo>, ClusterStateListener {
 
-    DocTableInfo getDropableTable(TableIdent tableIdent);
+    public static final Pattern SCHEMA_PATTERN = Pattern.compile("^([^.]+)\\.(.+)");
+    public static final String DEFAULT_SCHEMA_NAME = "doc";
 
-    DocTableInfo getWritableTable(TableIdent tableIdent);
+    private static final ESLogger LOGGER = Loggers.getLogger(Schemas.class);
```
Comment (dobe) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: use logger

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 4]
```diff
@@ -21,25 +21,244 @@
...
+                   DocSchemaInfoFactory docSchemaInfoFactory) {
+        super(settings);
+        this.clusterService = clusterService;
+        this.docSchemaInfoFactory = docSchemaInfoFactory;
+        schemas.putAll(builtInSchemas);
+        this.builtInSchemas = builtInSchemas;
+    }
 
-    boolean isOrphanedAlias(DocTableInfo table);
+    public DocTableInfo getDropableTable(TableIdent tableIdent) {
```
Comment (dobe) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: droppable

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 5]
```diff
@@ -21,25 +21,244 @@
...
+                try {
+                    schemas.remove(deletedSchema).close();
+                } catch (Exception e) {
+                    LOGGER.error(e.getMessage(), e);
+                }
+            }
+
+            // trigger changed event on all existing schemas
+            for (SchemaInfo schemaInfo : this) {
+                schemaInfo.clusterChanged(event);
```
Comment (dobe) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: maybe rename that clusterChanged method to something like update, in order to not confuse it with the listener method.

File: app/src/test/java/io/crate/node/NodeSettingsTest.java [hunk 6]
```diff
@@ -164,4 +160,24 @@ public void testDefaultPorts() throws IOException {
...
+    public void testInvalidUnicastHost() throws Exception {
+        Settings.Builder builder = Settings.settingsBuilder()
+            .put("discovery.zen.ping.multicast.enabled", false)
+            .put("discovery.zen.ping.unicast.hosts", "nonexistinghost:4300")
+            .put("path.home", CRATE_CONFIG_PATH);
+        Terminal terminal = Terminal.DEFAULT;
+        Environment environment = CrateSettingsPreparer.prepareEnvironment(builder.build(), terminal);
+
+        try {
+            node = new CrateNode(environment);
```
Comment (seut) [inline app/src/test/java/io/crate/node/NodeSettingsTest.java]: afaik this complex exception checking is not possible using the exception rule

Comment (dobe) [inline app/src/test/java/io/crate/node/NodeSettingsTest.java]: ok ... but still pls assert that it does fail

Comment (seut) [inline app/src/test/java/io/crate/node/NodeSettingsTest.java]: ah sure, thx

File: sql/src/main/java/io/crate/metadata/Schemas.java [hunk 7]
```diff
@@ -253,6 +253,7 @@ protected void doStart() {
 
     @Override
     protected void doStop() {
+        clusterService.remove(this);
```
Comment (dobe) [inline sql/src/main/java/io/crate/metadata/Schemas.java]: if this is required, the you should do this in the other services too which do add(this)

**Pull request merged**